ur_robot_driver/src/ros/hardware_interface.cpp

```diff
@@ -628,6 +657,45 @@ void HardwareInterface::publishPose()
   }
 }
 
+void HardwareInterface::extractRobotStatus()
+{
+  using namespace rtde_interface;
+
+  robot_status_resource_.mode = robot_status_bits_[toUnderlying(UrRtdeRobotStatusBits::IS_TEACH_BUTTON_PRESSED)] ?
+                                    RobotMode::MANUAL :
+                                    RobotMode::AUTO;
+
+  robot_status_resource_.e_stopped = safety_status_bits_[toUnderlying(UrRtdeSafetyStatusBits::IS_EMERGENCY_STOPPED)] ?
+                                         TriState::TRUE :
+                                         TriState::FALSE;
+
+  // Note that this is true as soon as the drives are powered,
+  // even if the brakes are still closed
+  // which is in slight contrast to the comments in the
+  // message definition
+  robot_status_resource_.drives_powered =
+      robot_status_bits_[toUnderlying(UrRtdeRobotStatusBits::IS_POWER_ON)] ? TriState::TRUE : TriState::FALSE;
+
+  robot_status_resource_.motion_possible =
```

Comment:
I think a union would be good here, as well. Best achieved by combination with the following safety status bits:
```
  IS_PROTECTIVE_STOPPED = 2,
  IS_RECOVERY_MODE = 3,
  IS_SAFEGUARD_STOPPED = 4,
  IS_SYSTEM_EMERGENCY_STOPPED = 5,
  IS_ROBOT_EMERGENCY_STOPPED = 6,
  IS_EMERGENCY_STOPPED = 7,
  IS_VIOLATION = 8,
  IS_FAULT = 9,
  IS_STOPPED_DUE_TO_SAFETY = 10
```

Sorry for coming with that this late, it just came up while testing.

Reply:
Can't we check `in_error`? Are there any bits in this list which are not in the other?

Reply:
I did not propose `IS_SAFEGUARD_STOPPED` for `in_error` as I wouldn't consider that an error status. However, motion will not be possible, while this is active...

Reply:
Hm. Other industrial robots will report this as an error I believe.

If we don't, then we can run into the situation where `in_error` is `false` and `motors_on` is `true`, but `motion_possible` is `false`. That would at least be confusing.

Reply:
I thought, this would be exactly the purpose of this field. Otherwise it would be redundant, right?

But if that's the way this is usually being handled, I am absolutely fine with it.

Reply:
Anything which is not an error but would prevent motion should be used to determine the value of `in_motion`.

The two are not necessarily identical. Consider fi a Fanuc/KUKA/ABB/etc: controller in manual mode, TP in the hands of the operator. Operator holds the deadman switch(es), resets the controller, but does not turn on power -> `motion_possible` == `false`.

Enables drive power: `motion_possible` == `true`.

Keeps drive power enabled, but lets go of the deadman: `motion_possible` == `false`.

And in case there is a driver involved: no errors on the controller, drives powered, but the motion task of the driver is not active: `motion_possible` == `false`.

Reply:
This is exactly how I understood that field. And this is exactly my argument here, as well. I consider the `IS_SAFEGUARD_STOPPED` not an error. From my understanding this is used when an external device is connected to the controller that prevents motion once triggered, e.g. a door sensor, a laser scanner or a floor matt.

Reply:
Then `motion_possible` should be (pseudo):

```python
in_motion = in_error or (IS_SAFEGUARD_STOPPED in robot_status_bits_)
```

I believe.

Reply:
Alright, I will modify the expression for motion_possible shortly. However I don't have time to include the status of the external control program, this condition may be added later if that's okay.

Reply:
Yes, absolutely!

Reply:
> ```python
> motion_possible = in_error or (IS_SAFEGUARD_STOPPED in robot_status_bits_)
> ```
> 

I guess you mean the negated version of this expression. Motion is **not** possible if controller is in error or safeguard stop

Reply:
Yes, of course.

Updated.